dev-notes/minimal-os-for-fobnail.md

```diff
@@ -0,0 +1,407 @@
+
+## Different OSs propositions
+
+The research effect is presented below. 4 systems were considered:
+
+* Zephyr
+* Xous
+* seL4
+* Linux
+
+Each of them has a short description, an analysis of the launch in DLME, and the
+possibilities and potential problems that will have to be addressed for the
+Fobnail Token to be functional.
+
+### Zephyr
+
+Zephyr is a well-known and supported embedded RTOS with a monolithic kernel
+written in C. It supports various devices ranging from Cortex-M to x86 CPUs.
+It's main benefits are IPv4, CoAP, and USB (including USB EEM) support which are
+required for communication with Fobnail token. Other benefits are listed
+[here](https://www.zephyrproject.org/benefits/).
+
+#### Running in DLME
+
+Secure launch uses a component called
+[SKL](https://github.com/TrenchBoot/secure-kernel-loader) (Secure Kernel Loader)
+which is responsible for booting OS in DLME (Zephyr in that case). Zephyr boots
+using Multiboot1 but SKL does not support it (only Multiboot2 is supported). For
+SKL to be able to boot Zephyr, either Zephyr would have to support Multiboot2 or
+SKL Multiboot1.
+
+Also, Zephyr may have trouble with running on different hardware configurations.
+Some important HW-related configuration is baked into Zephyr during build:
+
+- LAPIC base is selected by `CONFIG_LOAPIC_BASE_ADDRESS`, if target platform has
+  LAPIC remapped using `IA32_APIC_BASE` Zephyr will break.
+
+- APIC mode (xAPIC vs x2APIC) is selected when building Zephyr and cannot be
+  detected at runtime. Zephyr may not boot if it has different APIC mode set
+  than firmware has.
+
+#### Fobnail integration
+
+Zephyr provides most of the drivers needed for integration:
+
+- USB hub drivers
+- USB EEM driver
+- Network stack with IPv4 and CoAP support
+
+TPM driver is missing, however there is a fairly recent
+[PoC implementation](https://github.com/drandreas/zephyr-tpm2-poc) of TPM2
+stack.
+
+Zephyr provides good support for standard C library (newlib), which should
+simplify [fobnail-attester](https://github.com/fobnail/fobnail-attester)
+porting.
+
+### Xous
+
+Xous is a microkernel-based OS written in Rust, currently in Alpha state. Xous
+is used as a secure microkernel for [Betrusted](betrusted.io) device and is
+focused on RISC-V, which right now is the only architecture supported. However,
+x86 already contains
+[dummy implementation](https://github.com/betrusted-io/xous-core/blob/30b82b25b100e958790973c129dc49e1acca79ec/kernel/src/arch/x86_64.rs)
+which probably will be extended someday.
+
+
+#### Running in DLME
+
+Xous currently cannot run on x86, and DLME on RISC-V is not covered here. We
+have opened an [issue](https://github.com/betrusted-io/xous-core/issues/139)
+raising the subject, among others, about x86 support.
+
+#### Fobnail integration
+
+Xous has network support but currently lacks USB host drivers and TPM support.
+Also, it lacks C interface for calling the kernel and lacks C library. Getting
```

Comment:
This is not a serious problem because one can either write the userland code in Rust (preferred) or write C bindings to the Rust code.

Reply:
@DemiMarie Thank you, we've made some corrections. Can you please resolve this thread?

Reply:
I allowed myself to resolve this thread as it contained a minor change